src/preset.ts

```diff
@@ -988,12 +997,131 @@ async function expandConfigurePresetImpl(folder: string, name: string, workspace
+                    // If no toolset version specified then choose the latest VS instance for the given generator
+                    vsInstall = vs;
+                    break;
+                }
+            }
+        }
+    }
+
+    if (!vsInstall) {
+        // I don't think we want to do this anymore, or at least it's conditional on the mode being "auto". If it's "always", we just want to report that we couldn't find a vs isntallation.
+        // I think this is fine for when we've done it in auto, but when we
+        if (opts.compilerName) {
+            log.error(localize('specified.cl.not.found',
+                "Configure preset {0}: Compiler {1} with toolset {2} and architecture {3} was not found, you may need to run the 'CMake: Scan for Compilers' command if this toolset exists on your computer.",
+                opts.preset.name, `"${opts.compilerName}.exe"`, toolset.version ? `"${toolset.version},${toolset.host}"` : `"${toolset.host}"`, `"${arch}"`));
+        } else {
+            log.error(localize('vs.not.found', "Configure preset {0}: No Visual Studio installation found that supports the specified toolset {1} and architecture {2}, you may need to run the 'CMake: Scan for Compilers' command if this toolset exists on your computer.",
+                opts.preset.name, toolset.version ? `"${toolset.version},${toolset.host}"` : `"${toolset.host}"`, `"${arch}"`));
+        }
+    } else {
+        log.info(localize('using.vs.instance', "Using developer environment from Visual Studio (instance {0}, version {1}, installed at {2})", vsInstall.instanceId, vsInstall.installationVersion, `"${vsInstall.installationPath}"`));
+        const vsEnv = await varsForVSInstallation(vsInstall, toolset.host!, arch, toolset.version);
+        const compilerEnv = vsEnv ?? EnvironmentUtils.create();
+
+        if (opts.shouldInterrogateForNinja) {
+            const vsCMakePaths = await paths.vsCMakePaths(vsInstall.instanceId);
+            if (vsCMakePaths.ninja) {
+                log.warning(localize('ninja.not.set', 'Ninja is not set on PATH, trying to use {0}', vsCMakePaths.ninja));
+                compilerEnv['PATH'] = `${path.dirname(vsCMakePaths.ninja)};${compilerEnv['PATH']}`;
+            }
+        }
+
+        return compilerEnv;
+    }
+}
+
+/**
+ * This method tries to apply, based on the useVsDeveloperEnvironment setting value and, in "auto" mode, whether certain preset compilers/generators are used and not found, the VS Dev Env.
+ * @param preset Preset to modify the parentEnvironment of. If the developer environment should be applied, the preset.environment is modified by reference.
+ * @param workspaceFolder The workspace folder of the CMake project.
+ * @param sourceDir The source dir of the CMake project.
+ * @returns Void. We don't return as we are modifying the preset by reference.
+ */
+async function tryApplyVsDevEnv(preset: ConfigurePreset, workspaceFolder: string, sourceDir: string): Promise<void> {
```

Comment:
we used to check if `preset.__vsDevEnvApplied` before doing this (to see if we've already applied the env??) do we still want to check `__parentEnvironment`?

Reply:
Yes, checking `__parentEnvironment` prohibits us from rechecking for the devenv (and thus re-invoking where.exe and more) when we have to expand the configure preset for all of the other presets (build, test, etc).